cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/replication/AttachmentPullProcessor.java

```diff
@@ -0,0 +1,46 @@
+package com.cloudant.sync.replication;
+
+import com.cloudant.mazha.CouchClient;
+import com.cloudant.sync.datastore.Attachment;
+import com.cloudant.sync.datastore.AttachmentException;
+import com.cloudant.sync.datastore.PreparedAttachment;
+import com.cloudant.sync.datastore.UnsavedStreamAttachment;
+
+import java.io.InputStream;
+
+public class AttachmentPullProcessor implements CouchClient
+        .InputStreamProcessor<PreparedAttachment> {
+
+    private final DatastoreWrapper datastoreWrapper;
+    private final String name;
+    private final String contentType;
+    private final Attachment.Encoding encoding;
+    private final long length;
+    private final long encodedLength;
+
+    AttachmentPullProcessor(DatastoreWrapper wrapper, String name, String contentType, String
+            encoding, long length, long encodedLength) {
+        this.datastoreWrapper = wrapper;
+        this.name = name;
+        this.contentType = contentType;
+        this.encoding = Attachment.getEncodingFromString(encoding);
+        this.length = length;
+        this.encodedLength = encodedLength;
+    }
+
+    static PreparedAttachment pulLAttachmentWithRetry(CouchClient client, DatastoreWrapper
+            wrapper, String id, String rev, String name, String contentType, String encoding,
+                                                      long length, long encodedLength) {
+        return client.processAttachmentStream(id, rev, name, true, new AttachmentPullProcessor
+                (wrapper, name, contentType, encoding, length, encodedLength));
+    }
+
+    @Override
+    public PreparedAttachment processStream(InputStream stream) throws AttachmentException {
+        UnsavedStreamAttachment usa = new UnsavedStreamAttachment(stream, name, contentType,
+                encoding);
+        PreparedAttachment attachment = datastoreWrapper.prepareAttachment(usa, length,
```

Comment:
Yes, the coupling is awkward, but as soon as we leave the scope of the `processStream` method then we lose the scope of the retry. So if we return the `UnsavedStreamAttachment` here we are back to where we were and the call to `prepareAttachment` (which is what actually consumes the response stream) will happen without a retry.

I think the alternative is another sub-type of `UnsavedStreamAttachment` that either stores the bytes in memory or in another temporary file on disk so they can be consumed separately. I felt that the memory consumption was likely undesirable and since the `PreparedAttachment` already places the file temporarily on disk it seemed wasteful to do that operation twice.

I guess there might be a contrivance where the same temporary file on disk is simultaneously used by the `UnsavedStreamAttachment` and the `PreparedAttachment`, but the awkwardness there was the use of the datastore's own attachments directory, which necessitates some knowledge of the datastore and so only moves the coupling slightly.

Happy to see an example if there is a better way that I'm missing.